### PR TITLE
preserve integers decoding raw JSON values

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug

**Which issue(s) this PR fixes**:
* Fixes setting a default for a replicas field in a custom resource definition defining a scale subresource
* Fixes client-side validation to allow whole number values (e.g. `0`) for a field defined as a number (float)

- [x] fix kube-openapi number/integer validation and add tests
- [x] update kube-openapi dependency
- [x] add tests for CRD scale replica defaulting

**Does this PR introduce a user-facing change?**:
```release-note
Fixes a bug defining a default value for a replicas field in a custom resource definition that has the scale subresource enabled
```

/cc @sttts 
/sig api-machinery

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/json/json.go [hunk 1]
```diff
@@ -66,11 +66,36 @@ func Unmarshal(data []byte, v interface{}) error {
...
+	var err error
+	switch v2 := (*v).(type) {
+	case json.Number:
+		*v, err = convertNumber(v2)
+	case map[string]interface{}:
+		err = convertMapNumbers(v2, depth+1)
+	case []interface{}:
+		err = convertSliceNumbers(v2, depth+1)
+	}
+	return err
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/util/json/json.go:96]: would fall through. But would make it explicit.

[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/json/json.go:96]: added TestUnmarshalNil just for you :)

**Comment by liggitt** [comment]
flake https://github.com/kubernetes/kubernetes/issues/89719

/retest

**Comment by liggitt** [comment]
/priority important-soon

crd scale replicas default test added

File: staging/src/k8s.io/apimachinery/pkg/util/json/json.go [hunk 2]
```diff
@@ -66,11 +66,36 @@ func Unmarshal(data []byte, v interface{}) error {
 		// If the decode succeeds, post-process the map to convert json.Number objects to int64 or float64
 		return convertSliceNumbers(*v, 0)
 
+	case *interface{}:
+		// Build a decoder from the given data
+		decoder := json.NewDecoder(bytes.NewBuffer(data))
+		// Preserve numbers, rather than casting to float64 automatically
+		decoder.UseNumber()
+		// Run the decode
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/util/json/json.go:74]: nit: comments don't add much, just the second one

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/json/json.go:74]: was a copy/paste from the other stanzas, I don't feel that strongly

**Comment by sttts** [comment]
Looks good, just a nit.

/lgtm

File: staging/src/k8s.io/apimachinery/pkg/util/json/json_test.go [hunk 3]
```diff
@@ -278,42 +280,139 @@ func TestEvaluateTypes(t *testing.T) {
...
-		data, ok := m["data"]
-		if !ok {
-			t.Errorf("%s: decoded object missing data key: %#v", tc.In, m)
-			continue
-		}
-		if !reflect.DeepEqual(tc.Data, data) {
-			t.Errorf("%s: expected\n\t%#v (%v), got\n\t%#v (%v)", tc.In, tc.Data, reflect.TypeOf(tc.Data), data, reflect.TypeOf(data))
-			continue
+	for i, tc := range testCases {
+		t.Run(fmt.Sprintf("%d_map", i), func(t *testing.T) {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/json/json_test.go:284]: this first test-case just got indented into a subtest, viewing without whitespace helps - https://github.com/kubernetes/kubernetes/pull/89833/files?w=1

**Pull request merged**